File: 백준/Gold/1967. 트리의 지름/트리의 지름.cc
#include<bits/stdc++.h>
using namespace std;
using pii = pair<int, int>;

vector<vector<pii>> e;

pii dfs(int cur, int pre){
    int node = cur, d = 0;
    for(auto [nxt, nd]: e[cur]){
        if(nxt == pre) continue;
        auto [nnode, nnd] = dfs(nxt, cur);
        nnd += nd;
        if(nnd > d){
            d = nnd;
            node = nnode;
        }
    }
    return {node, d};
}

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int n; cin >> n; e.resize(n+1);
    for(int i = 0; i < n-1; i++){
        int u, v, w; cin >> u >> v >> w;
        e[u].push_back({v, w});
        e[v].push_back({u, w});
    }
    cout << dfs(dfs(1, 0).first, 0).second;
    return 0;
} 

File: 백준/Gold/1967. 트리의 지름/트리의 지름.py
import sys 
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

sys.setrecursionlimit(10**5)

def main():
    n = int(input())
    e = [[] for _ in range(n+1)]
    for _ in range(n-1):
        u, v, w = map(int, input().split())
        e[u].append((v, w))
        e[v].append((u, w))
    def dfs(cur, pre):
        node, d = cur, 0
        for nxt, nd in e[cur]:
            if nxt == pre: continue
            nnode, nnd = dfs(nxt, cur)
            nnd += nd
            if nnd > d: d, node = nnd, nnode
        return node, d
    print(dfs(dfs(1, 0)[0], 0)[1])
    return
if __name__ == '__main__':
    main()